### PR TITLE
[RTC] Experience Fragment Component with Localization Logic

### LINKED ISSUE #558: [RTC] Experience Fragment Component with Localization Logic
As an author I would like to select an experience fragment (XF) on a page or on a template. The XF component will render the XF that fits the localization format of the requested page.

The XF component should comply with the following localization formats:

**1. No localization**

site structure:
/content/my-site/...

experience fragment structure:
/content/experience-fragments/my-site/…

requested page:
/content/my-site/men/t-shirts

configured XF:
/content/experience-fragments/my-site/.../xf

-> rendered XF:
/content/experience-fragments/my-site/.../xf

**2. Regional localization (Live Copy only)**

site structure:
/content/my-site/master/...
/content/my-site/paris/...  (live copy of master)
/content/my-site/lyon/...   (live copy of master)

experience fragment structure:
/content/experience-fragments/my-site/master/...
/content/experience-fragments/my-site/paris/...  (live copy of master)
/content/experience-fragments/my-site/lyon/...   (live copy of master)

requested page:
/content/my-site/lyon/.../foo/bar

configured XF:
/content/experience-fragments/my-site/master/.../xf

-> rendered XF:
/content/experience-fragments/my-site/lyon/.../xf

**3. Language localization (Language Copy only)**

site structure:
/content/my-site/en/...
/content/my-site/fr/... (language copy of en)
/content/my-site/sp/... (language copy of en)

experience fragment structure:
/content/experience-fragments/my-site/en/...
/content/experience-fragments/my-site/fr/... (language copy of en)
/content/experience-fragments/my-site/sp/... (language copy of en)

requested page:
/content/my-site/fr/.../foo/bar

configured XF:
/content/experience-fragments/my-site/en/.../xf

-> rendered XF:
/content/experience-fragments/my-site/fr/…/xf

Note: the locale could also have the following format: ca-fr or ca_fr

**4. Multi-country localization**

site structure:
/content/my-site/master/en/...
/content/my-site/master/fr/... (language copy of master/en)
/content/my-site/master/sp/... (language copy of master/en)
/content/my-site/us/en/... (live copy of master/en)
/content/my-site/us/sp/... (live copy of master/sp)
/content/my-site/ca/en/... (live copy of master/en)
/content/my-site/ca/fr/... (live copy of master/fr)

experience fragment structure:
/content/experience-fragments/my-site/master/en/...
/content/experience-fragments/my-site/master/fr/... (language copy of master/en)
/content/experience-fragments/my-site/master/sp/... (language copy of master/en)
/content/experience-fragments/my-site/us/en/... (live copy of master/en)
/content/experience-fragments/my-site/us/sp/... (live copy of master/sp)
/content/experience-fragments/my-site/ca/en/... (live copy of master/en)
/content/experience-fragments/my-site/ca/fr/... (live copy of master/fr)

requested page:
/content/my-site/ca/fr/.../foo/bar

configured XF:
/content/experience-fragments/my-site/master/en/…/xf
or
/content/experience-fragments/my-site/us/en/…/xf

-> rendered XF:
/content/experience-fragments/my-site/ca/fr/.../xf

**5. Multi-country localization (optional)**

site structure:
/content/master/my-site/en/...
/content/master/my-site/fr/... (language copy of master/my-site/en)
/content/master/my-site/sp/... (language copy of master/my-site/en)
/content/us/my-site/en/... (live copy of master/my-site/en)
/content/us/my-site/sp/... (live copy of master/my-site/sp)
/content/ca/my-site/en/... (live copy of master/my-site/en)
/content/ca/my-site/fr/... (live copy of master/my-site/fr)

experience fragment structure:
/content/experience-fragments/master/my-site/en/...
/content/experience-fragments/master/my-site/fr/... (language copy of master/my-site/en)
/content/experience-fragments/master/my-site/sp/... (language copy of master/my-site/en)
/content/experience-fragments/us/my-site/en/... (live copy of master/my-site/en)
/content/experience-fragments/us/my-site/sp/... (live copy of master/my-site/sp)
/content/experience-fragments/ca/my-site/en/... (live copy of master/my-site/en)
/content/experience-fragments/ca/my-site/fr/... (live copy of master/my-site/fr)

requested page:
/content/ca/my-site/fr/.../foo/bar

configured XF:
/content/experience-fragments/master/my-site/en/…/xf
or
/content/experience-fragments/us/my-site/en/…/xf

-> rendered XF:
/content/experience-fragments/ca/my-site/fr/.../xf

### PR DESCRIPTION
**Pull request by jckautzmann**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #558  
| Patch: Bug Fix?          |
| Minor: New Feature?      |
| Major: Breaking Change?  |
| Tests Added + Pass?      | Yes
| Documentation Provided   | Yes (code comments and or markdown)
| Any Dependency Changes?  |
| License                  | Apache License, Version 2.0

### CODE REVIEW

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ExperienceFragmentImpl.java [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+    @Override public String getExportedType() {
+        return request.getResource().getResourceType();
+    }
+
+    /**
+     * Returns the localized page of the experience fragment page, based on the content page and the localization properties.
+     *
+     * @param xfPage the experience fragment page
+     * @param sitePage the content page
+     * @return
```
Comment (chkeller) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ExperienceFragmentImpl.java]: Use ```javax.annotation.CheckForNull```
I advice to document the return value and the meaning of a ```NULL```
Though this is private, it helps anybody unfamiliar to the class to understand the meaning without reading all code

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ExperienceFragmentImpl.java [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+        return fragmentPath;
+    }
+
+    @Override
+    public String getLocalizedFragmentPath() {
+        String localizedFragmentPath = null;
+        Page xfPage = pageManager.getPage(fragmentPath);
+        if (xfPage != null) {
+            Page xfLocalizedPage = getXfLocalizedPage(xfPage, currentPage);
+            if (xfLocalizedPage != null) {
```
Comment (chkeller) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ExperienceFragmentImpl.java]:
```suggestion
            if (xfLocalizedPage != null && xfLocalizedPage.hasContent()) {
```

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ExperienceFragmentImpl.java [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+                localizedFragmentPath = xfLocalizedPage.getContentResource().getPath();
+            } else {
```
Comment (chkeller) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ExperienceFragmentImpl.java]:
```suggestion
            } else if (xfPage.hasContent){
```

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ExperienceFragmentImpl.java [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+    @Override
+    public String getFragmentPath() {
+        return fragmentPath;
+    }
+
+    @Override
+    public String getLocalizedFragmentPath() {
+        String localizedFragmentPath = null;
+        Page xfPage = pageManager.getPage(fragmentPath);
+        if (xfPage != null) {
```
Comment (chkeller) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ExperienceFragmentImpl.java]:
```suggestion
        if (xfPage != null || xfPage.hasContent()) {
```

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ExperienceFragmentImpl.java [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+     * 2. Language localization                  | /content/my-site/en/my-page       | /content/my-site  | 1                 | en
+     * 3. Country-language localization          | /content/my-site/us/en/my-page    | /content/my-site  | 2                 | us/en
+     * 4. Country-language localization (variant)| /content/us/my-site/en/my-page    | /content          | 3                 | us/my-site/en
+     *
+     * @param page the page that contains the localization string
+     * @param root the localization root resource
+     * @param depth the localization depth
+     * @return the localization string
+     */
+    private String getLocalizationString(@NotNull Page page, @NotNull Resource root, int depth) {
```
Comment (chkeller) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ExperienceFragmentImpl.java]:
```suggestion
    private String getLocalizationString(@NotNull String pagePath, @NotNull Resource root, int depth) {
```
It's more flexible to just take what is really needed for a method

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ExperienceFragmentImpl.java [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+        String xfLocalizationSrcPath = StringUtils.joinWith(PATH_DELIMITER, xfLocalizationRoot, xfPageLocalizationString); // e.g. /content/experience-fragments/my-site/us/en
+        String xfLocalizationDestPath = StringUtils.joinWith(PATH_DELIMITER, xfLocalizationRoot, sitePageLocalizationString); // e.g. /content/experience-fragments/my-site/us/es
+
+        String xfLocalizedPagePath = StringUtils.replace(xfPage.getPath(), xfLocalizationSrcPath, xfLocalizationDestPath, 1);
+        return pageManager.getPage(xfLocalizedPagePath);
+    }
+
+    /**
+     * Returns the localization string of the given page based on the localization root and depth.
+     *
```
Comment (chkeller) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ExperienceFragmentImpl.java]: In fact the method is ignorant of localization.
It simply makes path replacement.

I would advice to reflect this in its name.

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ExperienceFragmentImpl.java [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+
+    @Override
+    public String getFragmentPath() {
+        return fragmentPath;
+    }
+
+    @Override
+    public String getLocalizedFragmentPath() {
+        String localizedFragmentPath = null;
+        Page xfPage = pageManager.getPage(fragmentPath);
```
Comment (chkeller) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ExperienceFragmentImpl.java]: Isn't ```fragmentPath``` potentially ```NULL```

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ExperienceFragmentImpl.java [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+    private String fragmentPath;
+
+    private int localizationDepth;
+    private String localizationRoot;
+    private ResourceResolver resolver;
+    private PageManager pageManager;
+
+    @PostConstruct
+    private void initModel() {
+        localizationRoot = properties.get(ExperienceFragment.PN_LOCALIZATION_ROOT, currentStyle.get(ExperienceFragment.PN_LOCALIZATION_ROOT, String.class));
```
Comment (chkeller) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ExperienceFragmentImpl.java]: Can you assert that ```localizationRoot``` is always neither "" nor ```NULL```?

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ExperienceFragmentImpl.java [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+     * @param xfPage the experience fragment page
+     * @param sitePage the content page
+     * @return
+     */
+    private Page getXfLocalizedPage(@NotNull Page xfPage, @NotNull Page sitePage) {
+        Resource siteRoot = resolver.getResource(localizationRoot);
+        if (siteRoot == null) {
+            return null;
+        }
+        String xfLocalizationRoot = StringUtils.replace(localizationRoot, CONTENT_ROOT, EXPERIENCE_FRAGMENTS_ROOT, 1);
```
Comment (chkeller) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ExperienceFragmentImpl.java]: as CONTENT_ROOT endsWith() != "/" => /content-en/x  would be replaced to EXPERIENCE_FRAGMENT_ROOTS+ x

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ExperienceFragmentImpl.java [hunk 10]
```diff
@@ -0,0 +1,183 @@
+/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ ~ Copyright 2019 Adobe Systems Incorporated
+ ~
```
Comment (chkeller) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ExperienceFragmentImpl.java]: Adobe changed name to Adobe.
Along there is a new License header please replace

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/models/ExperienceFragment.java [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+     */
+    String PN_LOCALIZATION_DEPTH = "localizationDepth";
+
+    /**
+     * Returns the configured path of the experience fragment variation.
+     *
+     * @return Configured experience fragment variation path
+     * @since com.adobe.cq.wcm.core.components.models 12.9.0
+     */
+    default String getFragmentPath() {
```
Comment (chkeller) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/models/ExperienceFragment.java]: It's a matter of convention of this project.

My personal point of view:
It is inconsistent to declare a method on API level: Meaning It is a characteristics of the Interface
But providing a default that doesn't support this characteristics

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/models/ExperienceFragment.java [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+     * content tree relative to the localization root. The property should provide a Long value.
+     *
+     * @since com.adobe.cq.wcm.core.components.models 12.9.0
+     */
+    String PN_LOCALIZATION_ROOT = "localizationRoot";
+
+    /**
+     * Name of the configuration policy property that defines the localization root from which to build the global
+     * localization structure. The property should provide a String value.
+     *
```
Comment (chkeller) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/models/ExperienceFragment.java]: you mix-matched the java-doc with the one from localizationRoot

**Comment by jckautzmann** [comment]
I completely changed the logic in https://github.com/adobe/aem-core-wcm-components/pull/675/commits/7bce91c9a620b6d169695210a9d0e07347369a2e
The former implementation one was based on two configuration properties (localization root and depth) to deduct the localization structure. This implementation automatically detects the localization structure of the site by relying on the LanguageManager and the LiveRelationshipManager APIs.
The logic is only triggered for XF resources defined in a template.

New feedback is welcome!

cc: @chkeller

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ExperienceFragmentImpl.java [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+     * Checks if the resource exists at the given path.
+     *
+     * @param path The resource path.
+     * @return {@code true} if the resource exists, {@code false} otherwise.
+     */
+    private boolean resourceExists(String path) {
+        if (StringUtils.isEmpty(path)) {
+            return false;
+        }
+        return (resolver.getResource(path) != null);
```
Comment (lewandowskidawid) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ExperienceFragmentImpl.java]: could be simplified to `StringUtils.isNotEmpty(path) && resolver.getResource(path) != null`

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ExperienceFragmentImpl.java [hunk 14]
```diff
@@ -0,0 +1,221 @@
...
+                Page rootPage = languageManager.getLanguageRoot(resource);
+                if (rootPage != null) {
+                    // language root
+                    root = rootPage.getPath();
+                }
+                if (StringUtils.isEmpty(root)) {
+                    // no language root
+                    try {
+                        if (relationshipManager.isSource(resource)) {
+                            // the resource is a blueprint
```
Comment (lewandowskidawid) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ExperienceFragmentImpl.java]: you could replace comments with dedicated methods e.g.
`if (relationshipManager.isSource(resource))` -> `if(isBlueprintResource(resource))`
`(relationshipManager.hasLiveRelationship(resource)` -> `if(isLiveCopyResource(resource))`

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ExperienceFragmentImpl.java [hunk 15]
```diff
@@ -0,0 +1,221 @@
...
+     * 2. Language localization                  | /content/mysite/en/mypage            | /content/mysite/en
+     * 3. Country-language localization          | /content/mysite/us/en/mypage         | /content/mysite/us/en
+     * 4. Country-language localization (variant)| /content/us/mysite/en/mypage         | /content/us/mysite/en
+     * 5. Blueprint                              | /content/mysite/blueprint/mypage     | /content/us/mysite/blueprint
+     * 4. Livecopy                               | /content/mysite/livecopy/mypage      | /content/us/mysite/livecopy
+     *
+     * @param path the resource path.
+     * @return the localization root of the resource at the given path if it exists, {@code null} otherwise.
+     */
+    private String getRoot(String path) {
```
Comment (lewandowskidawid) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ExperienceFragmentImpl.java]: you might consider extraction of logic related to path calculation to separate class. the class might be enhanced about unit/integration tests which cover all cases

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ExperienceFragmentImpl.java [hunk 16]
```diff
@@ -0,0 +1,221 @@
...
+            String currentPageRootPath = getRoot(currentPage.getPath());
+            // we should use getRoot instead of getXfRoot once the XF UI supports creating Live and Language Copies
+            String xfRootPath = getXfRoot(fragmentPath, currentPageRootPath);
+            if (!StringUtils.isEmpty(currentPageRootPath) && !StringUtils.isEmpty(xfRootPath)) {
+                String xfRelativePath = StringUtils.substring(fragmentPath, xfRootPath.length());
+                String localizedXfRootPath = StringUtils.replace(currentPageRootPath, CONTENT_ROOT, EXPERIENCE_FRAGMENTS_ROOT, 1);
+                localizedFragmentPath = StringUtils.join(localizedXfRootPath, xfRelativePath, JCR_CONTENT_ROOT);
+            }
+        }
+        String xfContentPath = StringUtils.join(fragmentPath, JCR_CONTENT_ROOT);
```
Comment (lewandowskidawid) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ExperienceFragmentImpl.java]: should we consider a case when an XF is placed on a page level? if not the whole logic of that method could be placed inside `isTemplate` condition

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ExperienceFragmentImpl.java [hunk 17]
```diff
@@ -0,0 +1,221 @@
...
+    private LanguageManager languageManager;
+
+    @OSGiService
+    private LiveRelationshipManager relationshipManager;
+
+    private String localizedFragmentPath;
+
+    @PostConstruct
+    protected void initModel() {
+        if (inTemplate()) {
```
Comment (lewandowskidawid) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ExperienceFragmentImpl.java]: should we consider following cases:
an XF is placed on a page. Currently the view layer refers to `localizedFragmentPath` only. The `localizedFragmentPath` is calculated only in particular cases when an XF is placed on a template so in result nothing will be displayed
please correct me if I am wrong

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ExperienceFragmentImpl.java [hunk 18]
```diff
@@ -0,0 +1,221 @@
...
+    private String getXfRoot(String xfPath, String currentPageRoot) {
+        String xfRoot = null;
+        if (!StringUtils.isEmpty(xfPath) && !StringUtils.isEmpty(currentPageRoot)
+            && resolver.getResource(xfPath) != null && resolver.getResource(currentPageRoot) != null) {
+            String[] xfPathTokens = Text.explode(xfPath, PATH_DELIMITER_CHAR);
+            String[] referenceRootTokens = Text.explode(currentPageRoot, PATH_DELIMITER_CHAR);
+            int xfRootDepth = referenceRootTokens.length + 1;
+            if (xfPathTokens.length >= xfRootDepth) {
+                String[] xfRootTokens = new String[xfRootDepth];
+                System.arraycopy(xfPathTokens, 0, xfRootTokens, 0, xfRootDepth);
```
Comment (lewandowskidawid) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ExperienceFragmentImpl.java]: as alternative solution you could use `Arrays` class e.g.
`String[] xfRootTokens = Arrays.copyOf(xfPathTokens, xfRootDepth);`
or
`String[] xfRootTokens = Arrays.copyOfRange(xfPathTokens, 0, xfRootDepth);`
in my opinion the above approaches/methods are more descriptive and more intuitive

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ExperienceFragmentImpl.java (hunk 10 above)
Comment (jckautzmann) [reply bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ExperienceFragmentImpl.java]: Addressed in commit https://github.com/adobe/aem-core-wcm-components/commit/7bce91c9a620b6d169695210a9d0e07347369a2e

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ExperienceFragmentImpl.java (hunk 13 above)
Comment (jckautzmann) [reply bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ExperienceFragmentImpl.java]: Good catch! This was addressed in https://github.com/adobe/aem-core-wcm-components/pull/675/commits/7bce91c9a620b6d169695210a9d0e07347369a2e

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ExperienceFragmentImpl.java (hunk 14 above)
Comment (jckautzmann) [reply bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ExperienceFragmentImpl.java]: I would say that it's clear enough the was it is now.

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ExperienceFragmentImpl.java (hunk 15 above)
Comment (jckautzmann) [reply bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ExperienceFragmentImpl.java]: All cases should already be covered by JUnit tests.

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ExperienceFragmentImpl.java (hunk 18 above)
Comment (jckautzmann) [reply bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ExperienceFragmentImpl.java]: The way it is now might be a bit faster as `Arrays.copyOf`relies internally on `System.arraycopy`.
See more comments in https://stackoverflow.com/questions/2589741/what-is-more-efficient-system-arraycopy-or-arrays-copyof

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/models/ExperienceFragment.java [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+     * If both the content page and the experience fragment have a localized root (language, live copy or blueprint),
+     * - it is then assumed that the content pages and the experience fragments follow the same structure patterns -
+     * this method returns the localized path of the experience fragment based on the localization of the content page
+     * if it exists, otherwise it returns the fragment path that is configured, suffixed with "/jcr:content".
+     *
+     * @return Localized experience fragment variation path
+     * @since com.adobe.cq.wcm.core.components.models 12.9.0
+     */
+    default String getLocalizedFragmentPath() {
```
Comment (bpauli) [inline bundles/core/src/main/java/com/adobe/cq/wcm/core/components/models/ExperienceFragment.java]: Can we not just merge the getFragmentPath with getLocalizedFragmentPath? The description of the getLocalizedFragmentPath looks for me as a implementation detail.

**Comment by jckautzmann** [comment]
@bpauli I gave the test methods more meaningful names in https://github.com/adobe/aem-core-wcm-components/pull/675/commits/4d93570250feefba2c9d7320b1dd4aadc5e41e8c

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ExperienceFragmentImpl.java [hunk 20]
```diff
@@ -0,0 +1,218 @@
...
+                            LiveRelationship liveRelationship = relationshipManager.getLiveRelationship(resource, false);
+                            if (liveRelationship != null) {
+                                LiveCopy liveCopy = liveRelationship.getLiveCopy();
+                                if (liveCopy != null) {
+                                    root = liveCopy.getPath();
+                                }
+                            }
+                        }
+                    } catch (WCMException e) {
+                        // ignore
```
[the remark this answers is not in the document]
Comment (jckautzmann) [reply bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ExperienceFragmentImpl.java]: Addressed in https://github.com/adobe/aem-core-wcm-components/pull/675/commits/6331b0beb143635622df820658f207654454f828

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ExperienceFragmentImpl.java [hunk 21]
```diff
@@ -0,0 +1,218 @@
...
+     * 2. Language localization                  | /content/mysite/en/mypage            | /content/mysite/en
+     * 3. Country-language localization          | /content/mysite/us/en/mypage         | /content/mysite/us/en
+     * 4. Country-language localization (variant)| /content/us/mysite/en/mypage         | /content/us/mysite/en
+     * 5. Blueprint                              | /content/mysite/blueprint/mypage     | /content/us/mysite/blueprint
+     * 4. Livecopy                               | /content/mysite/livecopy/mypage      | /content/us/mysite/livecopy
+     *
+     * @param path the resource path.
+     * @return the localization root of the resource at the given path if it exists, {@code null} otherwise.
+     */
+    private String getRoot(String path) {
```
[the remark this answers is not in the document]
Comment (jckautzmann) [reply bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ExperienceFragmentImpl.java]: I addressed 1) and 2) in https://github.com/adobe/aem-core-wcm-components/pull/675/commits/6331b0beb143635622df820658f207654454f828.
For 3) the localization logic is only triggered when the XF in defined in a template, which limits the scope. For now I would leave the implementation as is. We can introduce some caching later if needed.

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/models/ExperienceFragment.java (hunk 19 above)
Comment (jckautzmann) [reply bundles/core/src/main/java/com/adobe/cq/wcm/core/components/models/ExperienceFragment.java]: I removed `getFragmentPath` in https://github.com/adobe/aem-core-wcm-components/pull/675/commits/2fa68ff1ef24f09e9eb5d9e0b7d125cbb96c460b

**Pull request merged**